### DISCUSSION TITLE
PBD Liquid diffuse particles custom addition for more control

### DISCUSSION THREAD

**Discussion by kresimirt** [NVIDIA-Omniverse/PhysX#297] [Ideas]
Hello,

We're using PBD liquids a lot and we have noticed that we cannot add new diffuse particles as system won't detect them even if number of active particles are increased by using the setMaxActiveDiffuseParticles. Particles are properly added to the system but PhysX doesn't pick them up. For a test  we have replaced existing particles and relocate them to custom positions so we can see that the issue is not on our side.

This is quite important feature that wold allow users to add foam particles at custom locations. One example if a glass of a sparkling water or champagne were user must be able to create bubbles on custom positions and let PhysX do the advection through the fluid. Currently we are not able to create such simulation by using the PhysX only.

Thank you.

**Comment by vreutskyy** [comment]
Hi @kresimirt 
I see setMaxActiveDiffuseParticles can't set a value higher than maxNumDiffuseParticles value you pass to PxgParticleAndDiffuseBuffer constructor. Have you increased that value too?

**Reply by kresimirt** [reply]
Hello @vreutskyy,

We have maxNumDiffuseParticles set to a very high number and for our test scene there is plenty of room between currently active diffuse particles and max diffuse particles. From our test it looks like setMaxActiveDiffuseParticles() doesn't increase number of active particles, so this could be a  bug?

Here are some details

status at the beginning of the frame: active 24076 max 2000000 newactive **24682**

..here we add out particles behind curently active particles
..call setMaxActiveDiffuseParticles with new increased particle count (24682), but getNbActiveDiffuseParticles() still returns 24076

end result: active **24076** max 2000000 newactive 24682

Thanks

**Reply by vreutskyy** [reply]
I see, we'll check it. Thanks for the report. It would help a lot if you could provide a minimal snippet to reproduce the issue. Thanks

**Reply by kresimirt** [reply]
This is part of a larger program so I can just send you a snippet related to the code that should add foam to the liquid.
positions is a vector of custom foam to be added and it's our own class, thus the conversion.
Hope this helps.

PxU32 numBufferDiffuseParticlesActive	= gParticleBuffer->getNbActiveDiffuseParticles();
PxU32 numBufferDiffuseParticlesMax	= gParticleBuffer->getMaxDiffuseParticles();
PxU32 totalParticlesFoam 		= numBufferDiffuseParticlesActive + positions.size();

totalParticlesFoam = std::min(totalParticlesFoam, numBufferDiffuseParticlesMax);

PxVec4* _gpudiffusePositions = gParticleBuffer->getDiffusePositionLifeTime();
if (!_gpudiffusePositions) return;

PxVec4* diffusePosLife = cudaContextManager->allocPinnedHostBuffer<PxVec4>(totalParticlesFoam);

	cudaContextManager->acquireContext();

		PxCudaContext* cudaContext = cudaContextManager->getCudaContext();
		cudaContext->memcpyDtoH(diffusePosLife, CUdeviceptr(_gpudiffusePositions), sizeof(PxVec4) * numBufferDiffuseParticlesActive);

		int cIndex(0);

		for (int i = numBufferDiffuseParticlesActive; i < totalParticlesFoam; ++i)
		{
			PxVec4 _pos;
			_pos.x = positions[cIndex].x;
			_pos.y = positions[cIndex].y;
			_pos.z = positions[cIndex].z;
			_pos.w =100.0f;

			diffusePosLife[i] = _pos;
			cIndex++;
		}

		gParticleBuffer->setMaxActiveDiffuseParticles(totalParticlesFoam);

		cudaContext = cudaContextManager->getCudaContext();
			cudaContext->memcpyHtoD(CUdeviceptr(_gpudiffusePositions), &diffusePosLife[0], sizeof(PxVec4) * totalParticlesFoam);
		cudaContextManager->releaseContext();

		cudaContextManager->freePinnedHostBuffer(diffusePosLife);

	cudaContextManager->releaseContext();

**Reply by kresimirt** [reply]
Hello,

Are there any news on this issue?
We've moved to v5.4.2-106.1 and this issue is still present.

Thank you.

**Reply by vreutskyy** [reply]
Hi @kresimirt 
No, not yet. It's on the to-do list, but we haven't gotten there so far.
I'll ping the guys. Will tell if there's news.

**Reply by kresimirt** [reply]
Thank you for the speedy response.

We've added PhysX PBD water to our FumeFX tool so I just want to share this clip which is all PhysX PBD liquid: 
https://youtu.be/Nu7KUVAT4nA

There are few issues with the PhysX PBD liquid where foam is usually spawn from spray particles and it can get really messy.
Also, the surface of the liquid is not very smooth and you can see some clumping-like effect, which we have later fixed with filtering in our meshing tool.

Anyway, great work nVidia side as it runs pretty smooth.
I hope that FLIP, MPM will come soon as well.

**Reply by adenzler-nvidia** [reply]
Hi @kresimirt 
What you are trying to accomplish is not supported by the current implementation of diffuse particles in PhysX, independent of what max number you are setting in the particle buffer.

Diffuse particles are generated by the simulation based on the pressure field and the velocities of the "normal" particles - the simulation is in total control and will spawn particles wherever it sees fit. So from the perspective of a user, they should be considered read-only. Additionally, they should be considered more of a visual effect rather than a simulation.

The maxActiveDiffuseParticles exists as a knob for the users to tune the effect without having to resize the underlying GPU buffer.

Nice job on the integraton into FumeFX. Looks amazing :)

**Reply by kresimirt** [reply]
Hello @adenzler-nvidia ,

Thank you for the clarification. We plan to write our own foam code anyways, but it would be nice to have ability for user to add diffuse particles to PhysX PBD liquids as well.

Is there maybe a roadmap when (and if) we're going to see FLIP ad MPM in PhysX?

Thank you!
